consumer.go

```diff
@@ -162,8 +187,70 @@ func (c *Consumer) run() {
 			}
 
 		case <-c.done:
+			log.Println("Consumer initiating shutdown sequence")
+			// Send a CLS to all Cmd Channels for all connections
 			c.close()
+			log.Println("awaiting connection waitgroup")
+			// Wait for all runConn routines to return
 			c.join.Wait()
+			// At this point all runConn routines have returned, therefore we know
+			// we won't be receiving and new messages from nsqd servers. Now we can
+			// begin the processes of draining any in-flight messages and issuing a
+			// REQ command for each message.
+			log.Println("draining and requeueing remaining in-flight messages")
+			// drain and requeue any in-flight messages
+			c.drainRemaining()
+			log.Println("closing and cleaning up connections")
+			// Cleanup remaining connections
+			c.mtx.Lock()
+			for addr, cm := range c.conns {
+				delete(c.conns, addr)
+				// At this point we have drained all Messages from our main msgs channel and
+				// sent them REQ commands for each on their associated CmdChan. However, we can not simply just
+				// close the CmdChan for each connection yet. These channels are buffered and if
+				// we simply call closeCommand(cm.CmdChan) here there is a race, as the writeConn routines
+				// may not have finished processing all the REQ commands.
+				// Therefore we check the length of the channel and await for it to reach 0. If for some reason
+				// it fails to drain after a number of attempts we continue on and allow the messages to simply timeout
+				// and be reqeueued by the nsqd server.
+				attempts := 0
+				for len(cm.CmdChan) > 0 {
+					// If we've tried to allow the messages to flush and they've failed after
+					// 3 attempts we give up and let the nsqd instance timeout and requeue for us.
+					if attempts > 2 {
+						log.Printf("failed to requeue %d messages during orderly shutdown, allow them to timeout and requeue", len(cm.CmdChan))
+						break
+					}
+					log.Println("awaiting for write channel to flush any requeue commands")
+					time.Sleep(c.drainTimeout / 3)
+					attempts++
+				}
+				closeCommand(cm.CmdChan)
+				cm.Con.Close()
+			}
+			c.mtx.Unlock()
+			log.Println("Consumer exiting run")
+			// Signal to the stop() function that orderly shutdown is complete
+			c.shutJoin.Done()
+			return
+		}
+	}
+}
+
+// drainRemaining takes any remaining in-flight messages from the Consumer.msgs
+// channel and issues a REQ command for each.
+func (c *Consumer) drainRemaining() {
+	for {
```

Comment:
I think it would be safe to close `c.msgs` once the first join has come back if you wanted to make this a simple `range` on the channel.